internal/tsetse/rules/equals_nan_rule.ts

```diff
@@ -0,0 +1,30 @@
+/**
+ * @fileoverview Bans `== NaN` in TypeScript code, since it is always false for
+ * any value.
+ */
+
+import * as tsutils from 'tsutils';
+import * as ts from 'typescript';
+
+import {Checker} from '../checker';
+import {AbstractRule} from '../rule';
+
+const FAILURE_STRING = 'x == NaN always returns false; use isNaN(x) instead';
+
+export class Rule extends AbstractRule {
+  readonly ruleName = 'equals-nan';
+  readonly code = 21222;
+
+  register(checker: Checker) {
+    checker.on(
+        ts.SyntaxKind.BinaryExpression, checkBinaryExpression, this.code);
+  }
+}
+
+function checkBinaryExpression(checker: Checker, node: ts.BinaryExpression) {
+  if ((node.operatorToken.kind === ts.SyntaxKind.EqualsEqualsToken ||
+       node.operatorToken.kind === ts.SyntaxKind.EqualsEqualsEqualsToken) &&
+      (node.left.getText() === 'NaN' || node.right.getText() === 'NaN')) {
```

Comment:
can we query the type checker for the actual NaN symbol rather than using the text?

I'm not sure it makes any difference though, I don't think you could get a node whose text is `NaN` other than with the real symbol.

Reply:
According to astexplorer, TS just treats it as an identifier, not a special cased symbol.

I was surprised to see that I can actually do `var NaN = 3;` in Chrome with no error (though it doesn't actually do anything).